### ISSUE TITLE
Feature Request : s3 as destination / source

### ISSUE THREAD

**Issue by hilt86** [taskrabbit/elasticsearch-dump#473]
Thanks so much for making elasticsearch-dump! I was wondering if it is possible to add support for Amazon S3 as a source or destination?

**Comment by ferronrsmith** [comment]
Can you provide a bit more info. I don't use amazon services so the request is a bit vague.

**Comment by hilt86** [comment]
Thanks for your reply @ferronrsmith. Amazon S3 stands for simple storage service (i.e S x 3) that presents a REST api to store files on Amazons servers.

There are clients for [node](https://github.com/andrewrk/node-s3-client) and[ Amazon's official documentation](https://docs.aws.amazon.com/AmazonS3/latest/API/Welcome.html)

**Comment by ferronrsmith** [comment]
should be simple to implement and very similar to the `file` transport. Feel free to submit a PR.

**Comment by hilt86** [comment]
Thanks @ferronrsmith. I guess you are suggesting that I implement the functionality and submit a pull request?

**Comment by ferronrsmith** [comment]
Should be simple to implement, but I don't have access to s3, which would be required to implement + test this feature.

**Comment by hilt86** [comment]
ok, I will create you a bucket and send you credentials with which to test - I can send them to you if you send me your email?

**Comment by ferronrsmith** [comment]
started working on this. Just got a backlog of work, I'll push this out by EOW.

**Comment by manugarri** [comment]
looking forward to this !

**Comment by ferronrsmith** [comment]
Sorry for the delay. I release a **beta** version of this on the `s3-export` branch. This release only supports exporting to s3, not importing.

**Comment by ferronrsmith** [comment]
please test and let me know ASAP.

**Comment by hilt86** [comment]
Hi @ferronrsmith 

I cloned the repo and ran
`npm install .`

then I ran a job:

`./bin/elasticdump --input=http://elastic:9200/ --type=data  --awsAccessKeyId blah --awsSecretAccessKey blah --s3Bucket blah --s3RecordKey test`


It seemed to be working:
`Tue, 06 Nov 2018 21:57:11 GMT | got 100 objects from source elasticsearch (offset: 12500)
Tue, 06 Nov 2018 21:57:11 GMT | sent 100 objects to destination s3, wrote 100`

yet when I checked there were no files in the bucket?

**Comment by ferronrsmith** [comment]
strange. it's working for me ...

./bin/elasticdump --input http://localhost:9200/apm-6.4.1-transaction-2018.10.24 --s3Bucket blah --awsAccessKeyId blah --awsSecretAccessKey blah

**Comment by ferronrsmith** [comment]
It suppose to say... File Upload at the end. If that doesn't happen then it's not done...

**Comment by hilt86** [comment]
If i use an authenticated elasticsearch instance:

`--input=https://blah:pass@balh.us-east-1.aws.found.io:9243/`

I get an error:

`Wed, 07 Nov 2018 02:06:39 GMT | Error Emitted => {"error":{"root_cause":[{"type":"security_exception","reason":"action [indices:data/read/search] requires authentication","header":{"WWW-Authenticate":"Basic realm=\"security\" charset=\"UTF-8\""}}],"type":"security_exception","reason":"action [indices:data/read/search] requires authentication","header":{"WWW-Authenticate":"Basic realm=\"security\" charset=\"UTF-8\""}},"status":401}`

perhaps it is trying to use the Aws creds to authenticate to Elasticsearch?

**Comment by manugarri** [comment]
hey @ferronrsmith , i see that you merged this, but is it usable? @hilt86 , did you get it running?

**Comment by ferronrsmith** [comment]
Yes we worked out the issues. Wouldn’t have merged h workable code

**Comment by manugarri** [comment]
nice, thanks @ferronrsmith will give it a try as well
